workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/selection/VaultWatcherService.kt

```diff
@@ -0,0 +1,143 @@
+        }
+
+    }
+
+
+    //owner -> tokenClass -> tokenIdentifier -> List
+    init {
+        vaultObserver?.initialValues?.forEach(::addTokenToCache)
+        vaultObserver?.source?.subscribe(::onVaultUpdate)
+    }
+
+    private fun onVaultUpdate(t: Vault.Update<FungibleToken<out TokenType>>) {
+        t.consumed.forEach(::removeTokenFromCache)
+        t.produced.forEach(::addTokenToCache)
+    }
+
+    private fun removeTokenFromCache(it: StateAndRef<FungibleToken<*>>) {
+        val idx = processToken(it.state.data)
+        val tokenSet = getTokenBucket(idx)
+        if (tokenSet.remove(it) != true) {
+            LOG.warn("Received a consumption event for a token ${it.ref} which was either not present in the token cache, or was not locked")
+        }
+    }
+
+    private fun addTokenToCache(stateAndRef: StateAndRef<FungibleToken<out TokenType>>) {
+        val token = stateAndRef.state.data
+        val (owner, type, typeId) = processToken(token)
+        val tokensForTypeInfo = getTokenBucket(owner, type, typeId)
+        val existingMark = tokensForTypeInfo.putIfAbsent(stateAndRef as StateAndRef<FungibleToken<TokenType>>, false)
+        existingMark?.let {
+            LOG.warn("Attempted to overwrite existing token ${stateAndRef.ref} during cache initialization, this suggests incorrect vault behaviours")
+        }
+    }
+
+    inline fun <T : TokenType> selectTokens(
+            owner: PublicKey,
+            amountRequested: Amount<IssuedTokenType<T>>,
+            predicate: ((StateAndRef<FungibleToken<TokenType>>) -> Boolean) = { true },
+            allowSubSelect: Boolean = false,
+            autoUnlockDelay: Duration = Duration.ofMinutes(5)
+    ): List<StateAndRef<FungibleToken<T>>> {
+        val set = getTokenBucket(owner, amountRequested.token.tokenType.tokenClass, amountRequested.token.tokenType.tokenIdentifier)
+        val lockedTokens = mutableListOf<StateAndRef<FungibleToken<TokenType>>>()
+        var amountLocked: Amount<IssuedTokenType<T>> = amountRequested.copy(quantity = 0)
+        for (tokenStateAndRef in set.keys) {
+            //does the token satisfy the (optional) predicate?
+            if (predicate.invoke(tokenStateAndRef)) {
+                //if so, race to lock the token, expected oldValue = false
+                if (set.replace(tokenStateAndRef, false, true)) {
+                    //we won the race to lock this token
+                    lockedTokens.add(tokenStateAndRef)
+                    val token = tokenStateAndRef.state.data
+                    amountLocked += token.amount as Amount<IssuedTokenType<T>>
+                    if (amountLocked >= amountRequested) {
+                        break
+                    }
+                }
+            }
+        }
+        if (!allowSubSelect && amountLocked < amountRequested) {
+            throw InsufficientBalanceException("Could not find enough tokens to satisfy token request")
+        }
+
+        unlocker.schedule({
```

Comment:
so if tokens dont get used they are auto unlocked after 5 minutes. Seems like a long time. I wonder if there is a clever way to lock tokens to a flow ID, then watch for when that flow ending successfully or with an error. If it's successful then we remove from set. If it's an error, then we unlock?